encoder.go

```diff
@@ -78,28 +82,65 @@ const (
 	AppsecFieldTagValueIgnore = "ignore"
 )
 
-type native interface {
-	int64 | uint64 | uintptr
+// WAFObject is the C struct that represents a WAF object. It is passed as-is to the C-world.
+// It is highly advised to use the methods on the object to manipulate it and not set the fields manually.
+type WAFObject = bindings.WAFObject
+
+// Encodable represent a type that can encode itself into a WAFObject.
+// The encodable is responsible for using the [pin.Pinner]
+// object passed in the [EncoderConfig] to pin the data referenced by the encoded [bindings.WAFObject].
+// The encoder must also use the [timer.Timer] passed in the [EncoderConfig] to
+// make sure it doesn't spend too much time doing its job.
+// The encoder must also respect the [EncoderConfig] limits and report truncations.
+type Encodable interface {
+	// Encode encodes itself as the WAFObject obj using the provided EncoderConfig and remaining depth allowed.
+	// It returns a map of truncation reasons and their respective actual sizes. If the error returned is not nil
+	// It is greatly advised to return errors from the waferrors package error when it matters.
```

Comment:
This feels like a mix of implementer notes & user documentation... I feel it'd be better to formulate everything as "user doc", while implying maintainer doc as a corrollary? Or vice-versa.

```suggestion
	// It returns a map of truncation reasons and their respective actual sizes. If the error returned is not nil,
	// it is greatly advised to return errors from the waferrors package error when it matters.
```